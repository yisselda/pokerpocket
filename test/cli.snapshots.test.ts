import { describe, it, expect } from 'vitest'
import { spawn } from 'child_process'
import { fileURLToPath } from 'url'
import { dirname, join } from 'path'

const __filename = fileURLToPath(import.meta.url)
const __dirname = dirname(__filename)
const CLI_PATH = join(__dirname, '../dist/cli.js')

function runCliCommand(
  commands: string[],
  options: { ascii?: boolean; timeout?: number } = {}
): Promise<{
  stdout: string
  stderr: string
  exitCode: number | null
}> {
  const { ascii = true, timeout = 5000 } = options

  return new Promise((resolve, reject) => {
    const env = { ...process.env }
    if (ascii) {
      env.DISABLE_UNICODE = '1' // Force ASCII mode for consistent snapshots
    }

    const child = spawn('node', [CLI_PATH], {
      stdio: ['pipe', 'pipe', 'pipe'],
      env,
    })

    let stdout = ''
    let stderr = ''
    let hasExited = false

    const timeoutId = setTimeout(() => {
      if (!hasExited) {
        child.kill('SIGTERM')
        reject(new Error(`CLI process timed out after ${timeout}ms`))
      }
    }, timeout)

    child.stdout?.on('data', data => {
      stdout += data.toString()
    })

    child.stderr?.on('data', data => {
      stderr += data.toString()
    })

    child.on('exit', code => {
      hasExited = true
      clearTimeout(timeoutId)
      resolve({ stdout, stderr, exitCode: code })
    })

    child.on('error', error => {
      hasExited = true
      clearTimeout(timeoutId)
      reject(error)
    })

    // Send commands
    if (child.stdin) {
      for (const cmd of commands) {
        child.stdin.write(cmd + '\n')
      }
      child.stdin.write('q\n') // Always quit
      child.stdin.end()
    }
  })
}

function normalizeOutput(output: string): string {
  // Remove ANSI color codes if any
  return (
    output
      // eslint-disable-next-line no-control-regex
      .replace(/\u001b\[[0-9;]*m/g, '')
      // Normalize line endings
      .replace(/\r\n/g, '\n')
      // Remove trailing whitespace from lines
      .split('\n')
      .map(line => line.trimEnd())
      .join('\n')
      .trim()
  )
}

describe('CLI Snapshot Tests', () => {
  it('help command output', async () => {
    const result = await runCliCommand(['help'])
    expect(result.exitCode).toBe(0)
    expect(result.stderr).toBe('')

    const normalizedOutput = normalizeOutput(result.stdout)
    expect(normalizedOutput).toMatchSnapshot()
  })

  it('players command output', async () => {
    const result = await runCliCommand(['players 3', 'status'])
    expect(result.exitCode).toBe(0)

    const normalizedOutput = normalizeOutput(result.stdout)
    expect(normalizedOutput).toMatchSnapshot()
  })

  it('seed command output', async () => {
    const result = await runCliCommand(['seed 123', 'status'])
    expect(result.exitCode).toBe(0)

    const normalizedOutput = normalizeOutput(result.stdout)
    expect(normalizedOutput).toMatchSnapshot()
  })

  it('deal command output', async () => {
    const result = await runCliCommand([
      'players 3',
      'seed 42',
      'deal',
      'status',
    ])
    expect(result.exitCode).toBe(0)

    const normalizedOutput = normalizeOutput(result.stdout)
    expect(normalizedOutput).toMatchSnapshot()
  })

  it('complete game sequence', async () => {
    const result = await runCliCommand([
      'players 3',
      'seed 123',
      'deal',
      'flop',
      'turn',
      'river',
      'showdown',
    ])
    expect(result.exitCode).toBe(0)

    const normalizedOutput = normalizeOutput(result.stdout)
    expect(normalizedOutput).toMatchSnapshot()
  })

  it('hole card reveal', async () => {
    const result = await runCliCommand([
      'players 3',
      'seed 999',
      'deal',
      'hole 1',
      'hole 2',
      'hole 3',
    ])
    expect(result.exitCode).toBe(0)

    const normalizedOutput = normalizeOutput(result.stdout)
    expect(normalizedOutput).toMatchSnapshot()
  })

  it('fold sequence', async () => {
    const result = await runCliCommand([
      'players 4',
      'seed 777',
      'deal',
      'fold 1',
      'fold 2',
      'status',
      'fold 3',
      'status',
    ])
    expect(result.exitCode).toBe(0)

    const normalizedOutput = normalizeOutput(result.stdout)
    expect(normalizedOutput).toMatchSnapshot()
  })

  it('error handling - invalid commands', async () => {
    const result = await runCliCommand([
      'invalid-command',
      'players',
      'players abc',
      'seed',
      'hole',
      'fold xyz',
    ])
    expect(result.exitCode).toBe(0)

    const normalizedOutput = normalizeOutput(result.stdout)
    expect(normalizedOutput).toMatchSnapshot()
  })

  it('error handling - out of range values', async () => {
    const result = await runCliCommand([
      'players 0',
      'players 100',
      'hole 0',
      'hole 999',
      'fold 0',
      'fold 100',
    ])
    expect(result.exitCode).toBe(0)

    const normalizedOutput = normalizeOutput(result.stdout)
    expect(normalizedOutput).toMatchSnapshot()
  })

  it('game state errors', async () => {
    const result = await runCliCommand([
      'flop', // Before deal
      'showdown', // Before deal
      'players 3',
      'seed 12345', // Use fixed seed for deterministic output
      'deal',
      'players 4', // During hand
      'flop',
      'flop', // Already flopped
      'showdown',
    ])
    expect(result.exitCode).toBe(0)

    const normalizedOutput = normalizeOutput(result.stdout)
    expect(normalizedOutput).toMatchSnapshot()
  })
})

describe('CLI Unicode vs ASCII Tests', () => {
  it('unicode mode output', async () => {
    const result = await runCliCommand(
      ['players 2', 'seed 42', 'deal', 'skipbet', 'flop'],
      { ascii: false }
    )
    expect(result.exitCode).toBe(0)

    const normalizedOutput = normalizeOutput(result.stdout)
    // Should contain unicode symbols
    expect(normalizedOutput).toMatch(/[♠♥♦♣]/u)
    expect(normalizedOutput).toMatchSnapshot()
  })

  it('ascii mode output', async () => {
    const result = await runCliCommand(
      ['players 2', 'seed 42', 'deal', 'skipbet', 'flop'],
      { ascii: true }
    )
    expect(result.exitCode).toBe(0)

    const normalizedOutput = normalizeOutput(result.stdout)
    // Should NOT contain unicode symbols, only ASCII
    expect(normalizedOutput).not.toMatch(/[♠♥♦♣]/u)
    expect(normalizedOutput).toMatch(/[shdc]/) // ASCII suit letters
    expect(normalizedOutput).toMatchSnapshot()
  })

  it('non-TTY defaults to ASCII', async () => {
    // This simulates piping output or redirecting, which should force ASCII
    const result = await runCliCommand(['players 2', 'seed 42', 'deal'], {
      ascii: true,
    })
    expect(result.exitCode).toBe(0)

    const normalizedOutput = normalizeOutput(result.stdout)
    expect(normalizedOutput).not.toMatch(/[♠♥♦♣]/u)
    expect(normalizedOutput).toMatch(/[shdc]/)
  })
})
