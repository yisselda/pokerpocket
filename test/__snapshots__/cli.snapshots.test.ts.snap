// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CLI Snapshot Tests > complete game sequence 1`] = `
"üÉè Poker Pocket CLI
Type "help" for commands
Players: 2, Phase: IDLE
Available: deal, players <n>

> Players set to 3
Players: 3, Phase: IDLE
Available: deal, players <n>

> Seed set to 123 for next deal
> Cards dealt!
Players: 3, Phase: PREFLOP
Hole: P1: 2 cards, P2: 2 cards, P3: 2 cards
Available: flop, fold <player>

> Flop dealt!
Players: 3, Phase: FLOP
Board: Ts 2d 7c
Hole: P1: 2 cards, P2: 2 cards, P3: 2 cards
Available: turn, showdown, fold <player>

> Turn dealt!
Players: 3, Phase: TURN
Board: Ts 2d 7c 9d
Hole: P1: 2 cards, P2: 2 cards, P3: 2 cards
Available: river, showdown, fold <player>

> River dealt!
Players: 3, Phase: RIVER
Board: Ts 2d 7c 9d Td
Hole: P1: 2 cards, P2: 2 cards, P3: 2 cards
Available: showdown, fold <player>

>
P1: 5c Kc  ‚áí  ONE PAIR (Kc,Ts,7c,9d,Td)
P2: Th Js  ‚áí  THREE OF A KIND (Th,Js,Ts,9d,Td)
P3: Ks Jc  ‚áí  ONE PAIR (Ks,Jc,Ts,9d,Td)

Winner(s): P2

Hand complete!

P1: 5c Kc  ‚áí  ONE PAIR (Kc,Ts,7c,9d,Td)
P2: Th Js  ‚áí  THREE OF A KIND (Th,Js,Ts,9d,Td)
P3: Ks Jc  ‚áí  ONE PAIR (Ks,Jc,Ts,9d,Td)

Winner(s): P2

Board: Ts 2d 7c 9d Td
Available: deal (new hand), players <n>

> Goodbye!"
`;

exports[`CLI Snapshot Tests > deal command output 1`] = `
"üÉè Poker Pocket CLI
Type "help" for commands
Players: 2, Phase: IDLE
Available: deal, players <n>

> Players set to 3
Players: 3, Phase: IDLE
Available: deal, players <n>

> Seed set to 42 for next deal
> Cards dealt!
Players: 3, Phase: PREFLOP
Hole: P1: 2 cards, P2: 2 cards, P3: 2 cards
Available: flop, fold <player>

> Players: 3, Phase: PREFLOP
Hole: P1: 2 cards, P2: 2 cards, P3: 2 cards
Available: flop, fold <player>

> Goodbye!"
`;

exports[`CLI Snapshot Tests > error handling - invalid commands 1`] = `
"üÉè Poker Pocket CLI
Type "help" for commands
Players: 2, Phase: IDLE
Available: deal, players <n>

> Unknown command: invalid-command
Type "help" for available commands

Quick commands: deal, flop, turn, river, showdown, status, players <n>
> Usage: players <number>
> Usage: players <number>
> Usage: seed <number>
> Usage: hole <player> (e.g. hole 1)
> Usage: fold <player> (e.g. fold 1)
> Goodbye!"
`;

exports[`CLI Snapshot Tests > error handling - out of range values 1`] = `
"üÉè Poker Pocket CLI
Type "help" for commands
Players: 2, Phase: IDLE
Available: deal, players <n>

> Error: Players must be between 2 and 9
> Error: Players must be between 2 and 9
> Player must be 1-9
> Player must be 1-9
> Player must be 1-9
> Player must be 1-9
> Goodbye!"
`;

exports[`CLI Snapshot Tests > fold sequence 1`] = `
"üÉè Poker Pocket CLI
Type "help" for commands
Players: 2, Phase: IDLE
Available: deal, players <n>

> Players set to 4
Players: 4, Phase: IDLE
Available: deal, players <n>

> Seed set to 777 for next deal
> Cards dealt!
Players: 4, Phase: PREFLOP
Hole: P1: 2 cards, P2: 2 cards, P3: 2 cards, P4: 2 cards
Available: flop, fold <player>

> P1 folds
Players: 4, Phase: PREFLOP
Hole: P1: FOLDED, P2: 2 cards, P3: 2 cards, P4: 2 cards
Folded: P1
Available: flop, fold <player>

> P2 folds
Players: 4, Phase: PREFLOP
Hole: P1: FOLDED, P2: FOLDED, P3: 2 cards, P4: 2 cards
Folded: P1, P2
Available: flop, fold <player>

> Players: 4, Phase: PREFLOP
Hole: P1: FOLDED, P2: FOLDED, P3: 2 cards, P4: 2 cards
Folded: P1, P2
Available: flop, fold <player>

> P3 folds

P4 wins! (All other players folded)

> Players: 4, Phase: PREFLOP
Hole: P1: FOLDED, P2: FOLDED, P3: FOLDED, P4: 2 cards
Folded: P1, P2, P3
Available: flop, fold <player>

> Goodbye!"
`;

exports[`CLI Snapshot Tests > game state errors 1`] = `
"üÉè Poker Pocket CLI
Type "help" for commands
Players: 2, Phase: IDLE
Available: deal, players <n>

> Error: Can only flop from preflop phase
> Error: Can only showdown from flop, turn, or river phase
> Players set to 3
Players: 3, Phase: IDLE
Available: deal, players <n>

> Seed set to 12345 for next deal
> Cards dealt!
Players: 3, Phase: PREFLOP
Hole: P1: 2 cards, P2: 2 cards, P3: 2 cards
Available: flop, fold <player>

> Error: Cannot change players mid-hand
> Flop dealt!
Players: 3, Phase: FLOP
Board: 5d 7h Ah
Hole: P1: 2 cards, P2: 2 cards, P3: 2 cards
Available: turn, showdown, fold <player>

> Error: Can only flop from preflop phase
>
P1: 9s 4h  ‚áí  HIGH CARD (9s,4h,5d,7h,Ah)
P2: Qh 8c  ‚áí  HIGH CARD (Qh,8c,5d,7h,Ah)
P3: 7c Qc  ‚áí  ONE PAIR (7c,Qc,5d,7h,Ah)

Winner(s): P3

Hand complete!

P1: 9s 4h  ‚áí  HIGH CARD (9s,4h,5d,7h,Ah)
P2: Qh 8c  ‚áí  HIGH CARD (Qh,8c,5d,7h,Ah)
P3: 7c Qc  ‚áí  ONE PAIR (7c,Qc,5d,7h,Ah)

Winner(s): P3

Board: 5d 7h Ah
Available: deal (new hand), players <n>

> Goodbye!"
`;

exports[`CLI Snapshot Tests > help command output 1`] = `
"üÉè Poker Pocket CLI
Type "help" for commands
Players: 2, Phase: IDLE
Available: deal, players <n>

>
Commands:
  deal              - Deal new hand
  flop              - Deal flop (3 community cards)
  turn              - Deal turn (4th community card)
  river             - Deal river (5th community card)
  showdown          - Evaluate hands and determine winner(s)
  players <n>       - Set number of players (2-9, IDLE only)
  hole <player>     - Show hole cards for player (1-based, e.g. hole 1)
  fold <player>     - Fold player's hand (1-based, e.g. fold 1)
  status            - Show current game state
  help              - Show this help
  q                 - Quit

Advanced:
  seed <n>          - Set RNG seed for reproducible games

> Goodbye!"
`;

exports[`CLI Snapshot Tests > hole card reveal 1`] = `
"üÉè Poker Pocket CLI
Type "help" for commands
Players: 2, Phase: IDLE
Available: deal, players <n>

> Players set to 3
Players: 3, Phase: IDLE
Available: deal, players <n>

> Seed set to 999 for next deal
> Cards dealt!
Players: 3, Phase: PREFLOP
Hole: P1: 2 cards, P2: 2 cards, P3: 2 cards
Available: flop, fold <player>

> P1: Qs Qc
> P2: 2s 9d
> P3: 9c 6s
> Goodbye!"
`;

exports[`CLI Snapshot Tests > players command output 1`] = `
"üÉè Poker Pocket CLI
Type "help" for commands
Players: 2, Phase: IDLE
Available: deal, players <n>

> Players set to 3
Players: 3, Phase: IDLE
Available: deal, players <n>

> Players: 3, Phase: IDLE
Available: deal, players <n>

> Goodbye!"
`;

exports[`CLI Snapshot Tests > seed command output 1`] = `
"üÉè Poker Pocket CLI
Type "help" for commands
Players: 2, Phase: IDLE
Available: deal, players <n>

> Seed set to 123 for next deal
> Players: 2, Phase: IDLE
Available: deal, players <n>

> Goodbye!"
`;

exports[`CLI Unicode vs ASCII Tests > ascii mode output 1`] = `
"üÉè Poker Pocket CLI
Type "help" for commands
Players: 2, Phase: IDLE
Available: deal, players <n>

> Players set to 2
Players: 2, Phase: IDLE
Available: deal, players <n>

> Seed set to 42 for next deal
> Cards dealt!
Players: 2, Phase: PREFLOP
Hole: P1: 2 cards, P2: 2 cards
Available: flop, fold <player>

> Flop dealt!
Players: 2, Phase: FLOP
Board: 3h Td Qd
Hole: P1: 2 cards, P2: 2 cards
Available: turn, showdown, fold <player>

> Goodbye!"
`;

exports[`CLI Unicode vs ASCII Tests > unicode mode output 1`] = `
"üÉè Poker Pocket CLI
Type "help" for commands
Players: 2, Phase: IDLE
Available: deal, players <n>

> Players set to 2
Players: 2, Phase: IDLE
Available: deal, players <n>

> Seed set to 42 for next deal
> Cards dealt!
Players: 2, Phase: PREFLOP
Hole: P1: 2 cards, P2: 2 cards
Available: flop, fold <player>

> Flop dealt!
Players: 2, Phase: FLOP
Board: 3‚ù§ T‚ô¶ Q‚ô¶
Hole: P1: 2 cards, P2: 2 cards
Available: turn, showdown, fold <player>

> Goodbye!"
`;
